### PR TITLE
Run gateway as Core-Lightning plugin

### PR DESCRIPTION
**Pull request by justinmoon**
With this PR user can generate a lightning invoice and get paid in via lightning gateway.

Still need to:
- Figure out all the parameters for route hints. I think we can do that in future PRs.
- There's a weird [nix-related CI failure](https://github.com/fedimint/minimint/runs/7142539461?check_suite_focus=true) where it's having trouble validating the hash of cln-rpc
- We should update `cln-plugin` to allow plugin CLI arguments (`minimint-cfg` in our case) that don't have default values https://github.com/ElementsProject/lightning/pull/5369.
- This PR basically makes running the gateway mandatory. Previously you didn't have to run it because we always generated a fake node pubkey in `client.json` and it was never used so that didn't matter. So I just made it required for this PR because all the data structures assume the gateway is always there. https://github.com/fedimint/minimint/pull/200 makes it actually optional but that's a deeper change so I separated it.
- Should probably add a rust integration test. But I think it should only run when we're using the "real" fixtures because it's basically testing integration with core-lightning. Is that possible @jkitman?

### CODE REVIEW

File: client/client-lib/src/clients/user.rs [hunk 1]
```diff
@@ -343,6 +346,19 @@ impl UserClient {
...
+        let gateway_route_hint = RouteHint(vec![RouteHintHop {
+            src_node_id: gateway.node_pub_key,
+            short_channel_id: 8,
+            fees: RoutingFees {
+                base_msat: 0,
+                proportional_millionths: 0,
+            },
+            cltv_expiry_delta: (8 << 4) | 1,
+            htlc_minimum_msat: None,
+            htlc_maximum_msat: None,
```
Comment (justinmoon) [inline client/client-lib/src/clients/user.rs:359]: Need to figure out exactly what to put in these fields

File: ln-gateway/src/lib.rs [hunk 2]
```diff
@@ -1,75 +1,67 @@
...
-        let ln_client = cln_rpc::ClnRpc::new(ln_socket)
-            .await
-            .expect("connect to ln_socket");
-        let ln_client = Mutex::new(ln_client);
-
-        Self::new(Arc::new(federation_client), Box::new(ln_client)).await
-    }
-
-    pub async fn new(
+    pub async fn run(
```
Comment (justinmoon) [inline ln-gateway/src/lib.rs:30]: Not sure what to do with this API. Seems moving background fetch handle into this struct is good. And we need to pass the mutex to the background fetcher and webserver to avoid concurrency problems. But it's a little weird to pass it as argument here ...

Also just realized I still need to pass the SharedLnGateway to background_fetch ...

File: scripts/integrationtest.sh [hunk 3]
```diff
@@ -53,16 +59,14 @@ done
...
+
 # Run the Rust integration tests against the real Bitcoin / Lightning services
 export MINIMINT_TEST_REAL=1
 export MINIMINT_TEST_DIR=$TMP_DIR
 cargo test -p minimint-tests -- --test-threads=1
 
-# Generate federation, gateway and client config
-$BIN_DIR/configgen -- $CFG_DIR 4 4000 5000 1000 10000 100000 1000000 10000000
-LN1_PUB_KEY="$($LN1 getinfo | jq -r '.id')"
-$BIN_DIR/gw_configgen -- $CFG_DIR "$LN1_DIR/regtest/lightning-rpc" $LN1_PUB_KEY
```
Comment (justinmoon) [inline scripts/integrationtest.sh:64]: Gateway now generates config automatically

File: scripts/integrationtest.sh [hunk 4]
```diff
@@ -99,13 +103,11 @@ function await_block_sync() {
 }
 await_block_sync
 
-# Start LN gateway
-$BIN_DIR/ln_gateway $CFG_DIR &
```
Comment (justinmoon) [inline scripts/integrationtest.sh:103]: Core lightning now starts the gateway

File: client/client-lib/src/api.rs [hunk 5]
```diff
@@ -97,18 +95,19 @@ pub enum ApiError {
...
+    #[error("Core error: {0}")]
+    CoreError(#[from] CoreError),
 }
 
 impl ApiError {
-    /// Returns `true` if the error means that the queried coin output isn't ready yet but might
+    /// Returns `true` if the error means that the queried output isn't ready yet but might
     /// become ready later.
-    pub fn is_retryable_fetch_coins(&self) -> bool {
+    pub fn is_retryable(&self) -> bool {
```
Comment (justinmoon) [inline client/client-lib/src/api.rs:107]: I should split this change into separate PR

Comment (justinmoon) [reply client/client-lib/src/api.rs:107]: https://github.com/fedimint/minimint/pull/175

File: ln-gateway/src/lib.rs [hunk 6]
```diff
@@ -148,37 +146,77 @@ impl LnGateway {
...
     }
 }
 
 impl Drop for LnGateway {
     fn drop(&mut self) {
-        self.fetcher.abort();
-        assert!(futures::executor::block_on(&mut self.fetcher).is_err());
+        self.webserver.abort();
+        // FIXME: this panics ...
+        // assert!(futures::executor::block_on(&mut self.webserver).is_err());
```
Comment (justinmoon) [inline ln-gateway/src/lib.rs:219]: This assertion fails cc @elsirion. Not sure how to correctly do this cleanup ...

**Review by jkitman — DISMISSED** [review]
Regarding the integration tests, does this change any behavior that would need testing?

If not, perhaps you can just modify the `integrationtests/README.md` so that users can run the LN plugin locally when testing with the real fixtures.

Additionally `scripts/latencytest.sh` probably needs a small update.

**Comment by justinmoon** [comment]
> Regarding the integration tests, does this change any behavior that would need testing?

This doesn't change any behavior W.R.T. existing integration tests like`receive_lightning_payment_valid_preimage` and `receive_lightning_payment_invalid_preimage`. The current tests do a lot of work calling internal functions in the right order and everything. It would be nice if a new integration test just (1) generated invoice, (2) had "ln2" pay it, and (3) asserted that balances eventually updated as expected. A true "integration test".

> If not, perhaps you can just modify the `integrationtests/README.md` so that users can run the LN plugin locally when testing with the real fixtures.

One tricky thing here is that we'd need to pass the `--minimint-cfg` argument to the plugin which points to the directory with all the JSON configs and sled db files. AFAIK the integration tests just do everything in-memory?

Another tricky thing is that the plugin would really need to be reloadable per-test, because the federation details might change every test.

I think the best option here would be to add a `--minimint-testing` flag to the plugin which causes the plugin to run without running a gateway. Then add a new RPC command to the plugin to run a gateway based on JSON federation config pass as argument to this RPC command. This way the plugin could start before the integration tests have even started. Each test that needs to exercise the plugin can send over the federation configs used by that test. 

This could happen in a follow-on PR.

> Additionally `scripts/latencytest.sh` probably needs a small update.

I didn't even know about latencytest.sh! Will do.

**Comment by jkitman** [comment]
> This doesn't change any behavior W.R.T. existing integration tests like`receive_lightning_payment_valid_preimage` and `receive_lightning_payment_invalid_preimage`. The current tests do a lot of work calling internal functions in the right order and everything. It would be nice if a new integration test just (1) generated invoice, (2) had "ln2" pay it, and (3) asserted that balances eventually updated as expected. A true "integration test".

Perhaps it's sufficient that is handled via the shell scripts, although I guess we need to expose a new CLI command for that.

> One tricky thing here is that we'd need to pass the `--minimint-cfg` argument to the plugin which points to the directory with all the JSON configs and sled db files. AFAIK the integration tests just do everything in-memory?

Ah, yes the current tests only use an in-memory DB, although I was meaning to enable a real DB for when running with the real fixtures.

> Another tricky thing is that the plugin would really need to be reloadable per-test, because the federation details might change every test.
> I think the best option here would be to add a `--minimint-testing` flag to the plugin which causes the plugin to run without running a gateway. Then add a new RPC command to the plugin to run a gateway based on JSON federation config pass as argument to this RPC command. This way the plugin could start before the integration tests have even started. Each test that needs to exercise the plugin can send over the federation configs used by that test.

Agreed, tearing down the LN fixtures and funding a channel would make the tests take forever, much faster just to reload the configs.

File: ln-gateway/src/lib.rs [hunk 7]
```diff
@@ -148,37 +146,77 @@ impl LnGateway {
...
+                        if let Err(e) = federation_client.fetch_coins(out_point).await {
+                            debug!(error = %e, "Fetching coins failed");
+                        }
+                    }
+                })
+                .collect::<Vec<_>>();
+            futures::future::join_all(pending_fetches).await;
+
+            // Handle messages from webserver and plugin
+            while let Ok(msg) = self.receiver.try_recv() {
```
[the remark this answers is not in the document]
Comment (justinmoon) [reply ln-gateway/src/lib.rs:196]: @Maan2003 had a [PR](https://github.com/fedimint/minimint/pull/141) to make our database traits async, but abandoned it in favor of making the change change after #173 is merged.

**Review by elsirion — APPROVED** [review]
